### PR TITLE
Add ability to output emojis to chatbox

### PR DESCRIPTION
**Pull request by hjdarnel**
This adds a Menu Entry to the All tab right click, when Emojis configuration setting is true. 
image
image
image

### CODE REVIEW

**Comment by dekvall** [comment]
I don't think it makes much sense to add this in it's current state. Without being able to click on an emoji to append its tag to the input, this is more or less a built in wiki page. 

However, if we do want to add something like this, I think that it's preferable to have the tag displayed under the icon instead of as a hover text since having to move the mouse in order to see what to type would get annoying pretty quickly.

**Comment by hjdarnel** [comment]
We discussed in Discord making this a right click on chat buttons and would instead just emit the icons and triggers into chat with `Client::addChatMessage`, probably with a timeout and be removed after a couple of seconds like the log in/out messages. I will update this to reflect

**Comment by hjdarnel** [comment]
To continue this conversation here: which Chat tab should have the right click option? I'm preferential to `All`, `Game`, or each tab. I'm currently considering something like "Show Emojis" for the menu option verbiage?

**Comment by Nightfirecat** [comment]
I think adding it to the "All" tab makes the most sense to me, and maybe also to the active tab? I'd start with just adding it to the "All" tab, I think. And `Show emojis` sounds like a good menu action name. :+1:

**Comment by hjdarnel** [comment]
ok, I removed the navbar and panel in favor of leaving in the "emitting into chat" implementation. I don't like the config name very much. Any feedback on the formatting? It's hard to make it look too good because the font isn't monospaced. We could do something with `StringUtils.leftPad()` to make the output formatted similar char lengths but it still won't look like a table.

I updated PR description and title appropriately

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 1]
```diff
@@ -54,19 +77,52 @@
...
+		return configManager.getConfig(EmojiConfig.class);
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event) throws Exception
+	{
+		if (event.getGroup().equals(CONFIG_GROUP_KEY))
+		{
+			shutDown();
+			startUp();
```
Comment (dekvall) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: I don't think you need to do a full shutDown/startUp, isn't it enough to just remove the potentially emitted message when the config for emitting the message is set to false?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 2]
```diff
@@ -54,19 +77,52 @@
...
+	private static final Pattern TAG_REGEXP = Pattern.compile("<[^>]*>");
+	private static final Pattern WHITESPACE_REGEXP = Pattern.compile("[\\s\\u00A0]");
+	private static final String EMIT = "Show";
+	private static final String MENU_TARGET = "Emoji Key";
+	private static final int EMOJI_EMIT_DURATION = 20;
+	private static final int EMOJI_PER_LINE = 8;
+	/**
+	 * queue of temporary messages added to the client
+	 */
+	private final Deque<EmojiEmitMessage> emojiEmitMessages = new ArrayDeque<>();
```
Comment (dekvall) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: I don't think we need to have multiples of these, if the player chooses to emit emojis in quick succession imo it's fine to just remove the current emitted message and send a new one.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 3]
```diff
@@ -166,6 +222,134 @@ public void onOverheadTextChanged(final OverheadTextChanged event)
...
+		}
+
+		MenuEntry[] entries = client.getMenuEntries();
+		for (MenuEntry entry : entries)
+		{
+			if (entry.getOption().equals(EMIT) && entry.getTarget().equals(MENU_TARGET))
+			{
+				return;
+			}
+		}
```
Comment (dekvall) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Is this required? I think it should be enough to just check `event.getOption()`

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 4]
```diff
@@ -166,6 +222,134 @@ public void onOverheadTextChanged(final OverheadTextChanged event)
...
+	{
+		ChatMessageBuilder builder = new ChatMessageBuilder();
+		Emoji[] allEmoji = getAllEmoji();
+
+		for (int i = 0; i < allEmoji.length; i++)
+		{
+			Emoji emoji = allEmoji[i];
+			builder
+				.img(modIconsStart + emoji.ordinal())
+				.append("  " + Text.unescapeTags(emoji.trigger) + "  ");
```
Comment (dekvall) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Maybe something like `FontTypeFace#getTextWidth` can be used when aligning these(?)
The current formatting unfortunately looks really messy.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 1 above)
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Furthermore, it is very bad practice to directly call `startUp` and `shutDown`. If the code they contain should be accessed from multiple sources, put it into a separate function.

File: runelite-api/src/main/java/net/runelite/api/MenuAction.java [hunk 5]
```diff
@@ -272,6 +272,10 @@
 	 * Menu action for configuring runelite overlays.
 	 */
 	RUNELITE_OVERLAY_CONFIG(1502),
+	/**
+	 * Menu action for triggering Emoji output to chat.
+	 */
+	RUNELITE_EMOJI_OUTPUT(1503),
```
Comment (Nightfirecat) [inline runelite-api/src/main/java/net/runelite/api/MenuAction.java]: Don't add a new menu action type. Use `RUNELITE` instead and check the menu entry's text when clicked.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 6]
```diff
@@ -1,5 +1,6 @@
 /*
- * Copyright (c) 2019, Lotto <https://github.com/devLotto>
+ * Copyright (c) 2020, Lotto <https://github.com/devLotto>
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Don't update copyright date of other entries.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 7]
```diff
@@ -54,19 +77,52 @@
...
+	@Provides
+	EmojiConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(EmojiConfig.class);
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event) throws Exception
+	{
+		if (event.getGroup().equals(CONFIG_GROUP_KEY))
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: This should also check the config key so future configurations (if any) will not enter this block.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 8]
```diff
@@ -54,19 +77,52 @@
...
+		{
+			shutDown();
+			startUp();
+		}
+	}
+
 	@Override
-	protected void startUp()
+	protected void startUp() throws Exception
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Does `startUp` throw an exception now? Why is this added?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 9]
```diff
@@ -54,19 +77,52 @@
...
 	private int modIconsStart = -1;
 
+	@Provides
+	EmojiConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(EmojiConfig.class);
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event) throws Exception
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: What exception is this method expected to throw?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 10]
```diff
@@ -54,19 +77,52 @@
...
+	{
+		if (event.getGroup().equals(CONFIG_GROUP_KEY))
+		{
+			shutDown();
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Where is `shutDown` defined?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 11]
```diff
@@ -166,6 +222,134 @@ public void onOverheadTextChanged(final OverheadTextChanged event)
 		event.getActor().setOverheadText(updatedMessage);
 	}
 
+	@Subscribe
+	public void onMenuEntryAdded(MenuEntryAdded event)
+	{
+		int childId = WidgetInfo.TO_CHILD(event.getActionParam1());
+		if (!config.showEmitEmojiEntry() || childId != WidgetInfo.RESIZABLE_VIEWPORT_ALL_CHAT_TAB.getChildId())
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Why is this only comparing child IDs? And why is it comparing only to `RESIZABLE_VIEWPORT_ALL_CHAT_TAB`? What happened to checking `FIXED_VIEWPORT_ALL_CHAT_TAB`?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 7 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: is it enough to check `		if (event.getGroup().equals(CONFIG_GROUP_KEY) && event.getKey().equals("showEmitEmojiEntry"))
` or do we need to point to a field of the EmojiConfig class?

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: That's fine.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 11 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: The child Id of the All tab on fixed and resize is the same, so I didn't catch that. I'll check both verbosely.

Do we need to check the Group Id of the `MenuEntryAdded` too?

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 1 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Yes, we don't need to do anything on config toggle on, and on toggle off we can just remove the message. Good call. Thanks for the explanation too.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 11 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: @Nightfirecat how do we check the exact id of the widget instead of the child id? Can we do something like this? 

```
int widgetId = event.getActionParam1();
if (!config.showEmitEmojiEntry()
	|| widgetId != WidgetInfo.RESIZABLE_VIEWPORT_ALL_CHAT_TAB.getId()
	|| widgetId != WidgetInfo.FIXED_VIEWPORT_ALL_CHAT_TAB.getId())
{
	return;
}
```

I've tried this but I'm not sure if it's correct or not, it's not working but it's possible the `WidgetInfo.RESIZABLE_VIEWPORT_ALL_CHAT_TAB` and `FIXED_VIEWPORT_ALL_CHAT_TAB` are just wrong (since I've added them). Not sure which id in the widget inspector to use

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: That code looks correct to me. The widget inspector will show the group and child IDs in the tree on the left, and should display the `ID` as a row when selecting the widget.

On second thought, it will likely be easier to use the MenuManager class to manage adding these custom menu entries to these widgets.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/Emoji.java [hunk 12]
```diff
@@ -90,7 +90,7 @@
 
 	private static final Map<String, Emoji> emojiMap;
 
-	private final String trigger;
```
Comment (dekvall) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/Emoji.java]: Use `@Getter` instead

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 4 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Yup, I completely agree. I updated the PR description with the output as it stands with the latest changes.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 11 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: I've refactored to use the MenuManager helper. Let me know how it looks.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/Emoji.java (hunk 12 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/Emoji.java]: Thanks :)

File: runelite-client/src/test/java/net/runelite/client/util/TextTest.java [hunk 13]
```diff
@@ -42,4 +42,12 @@ public void removeTags()
 		assertEquals("Remove no tags", Text.removeTags("Remove no tags"));
 	}
 
+	@Test
+	public void unescapeTags()
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/util/TextTest.java]: This should include cases where tags are removed, since that is included as part of the behavior

File: runelite-client/src/main/java/net/runelite/client/util/Text.java [hunk 14]
```diff
@@ -150,6 +151,38 @@ else if (c != '\r')
 		return out.toString();
 	}
 
+	/**
+	 * Unescape a string for widgets, replacing &lt;lt&gt; and &lt;gt&gt; with their unescaped counterparts
+	 */
+	public static String unescapeTags(String str)
+	{
+		StringBuffer out = new StringBuffer(str.length());
```
[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/util/Text.java]: I copy pasted this from escapeTags and flipped some logic. I'll use StringBuilder instead.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		return false;
+	}
+
+	@ConfigItem(
+		position = 2,
+		keyName = "emojiEmitDuration",
+		name = "Message duration",
+		description = "Configure how long the emoji hint message stays in chat"
+	)
+	@Units(Units.TICKS)
```
[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java]: Yeah, I agree. I changed this.

**Comment by hjdarnel** [comment]
> Also, why is `icon.png` added?

I originally added this to be uniform with other nav bar icons, where `icon.png` refers to displayed icon. Since we're not doing that way anymore I removed it.

File: runelite-client/src/main/java/net/runelite/client/util/Text.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/util/Text.java]: I stand corrected, we need to use StringBuffer because `Matcher#appendReplacement(StringBuilder sb, String replacement)` only exists since Java 9

File: runelite-client/src/test/java/net/runelite/client/util/TextTest.java [hunk 16]
```diff
@@ -42,4 +42,16 @@ public void removeTags()
...
+	{
+		assertEquals("Test", Text.unescapeTags("Test"));
+		assertEquals("<:)", Text.unescapeTags("<lt>:)"));
+		assertEquals(">:)", Text.unescapeTags("<gt>:)"));
+		assertEquals("\n", Text.unescapeTags("<br>"));
+		assertEquals("<col=FFFFFF This is a very special message.</col> >:)", Text.unescapeTags("<col=FFFFFF This is a very special message.</col> <gt>:)"));
+		assertEquals("<col=FFFFFF This is a very special message.</col> <:)", Text.unescapeTags("<col=FFFFFF This is a very special message.</col> <lt>:)"));
+		assertEquals("<col=FFFFFF This is a very special message.</col> \n:)", Text.unescapeTags("<col=FFFFFF This is a very special message.</col> <br>:)"));
+		assertEquals("<col=FFFFFF This is a very special message.</col>Not so much.", Text.unescapeTags("<col=FFFFFF This is a very special message.</col>Not so much."));
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/util/TextTest.java]: I meant test a string with some tag present and observe that the tag would be preserved in the result. eg. `<col=FFFFFF>Colored text</col> and some other text`

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 17]
```diff
@@ -63,12 +82,77 @@
...
+	{
+		return configManager.getConfig(EmojiConfig.class);
+	}
+
+	/**
+	 * Remove tags, except for &lt;lt&gt; and &lt;gt&gt;
+	 *
+	 * @return String with tags removed
+	 */
+	private static String removeTags(String str)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: This method has still been moved compared to where it was originally. This should be appearing on what is currently L333 of this file. Look at the diff view of this PR to see what I mean. (down where you've added `timeoutEmojiMessages()`)

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 18]
```diff
@@ -63,12 +82,77 @@
...
+
+	static final String CONFIG_GROUP_KEY = "emoji";
+	private static final String EMIT = "Show";
+	private static final String MENU_TARGET = "Emoji Key";
+	private static final WidgetMenuOption FIXED_INVENTORY_ALL_TAB = new WidgetMenuOption(EMIT,
+		MENU_TARGET, WidgetInfo.FIXED_VIEWPORT_ALL_CHAT_TAB);
+	private static final WidgetMenuOption RESIZABLE_INVENTORY_ALL_TAB = new WidgetMenuOption(EMIT,
+		MENU_TARGET, WidgetInfo.RESIZABLE_VIEWPORT_ALL_CHAT_TAB);
+	private static final int EMOJI_PER_LINE = 7;
+	//	 Holds current emoji message, if any
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]:
```suggestion
	//	Holds current emoji message, if any
```

File: runelite-client/src/test/java/net/runelite/client/util/TextTest.java (hunk 16 above)
Comment (hjdarnel) [reply runelite-client/src/test/java/net/runelite/client/util/TextTest.java]: My bad, I didn't realize what I had added there.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java (hunk 17 above)
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Yeah, sorry, the diff was confusing me. _Best as I can tell_ this is put back now.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 19]
```diff
@@ -63,12 +82,52 @@
 	@Inject
 	private ChatMessageManager chatMessageManager;
 
+	@Inject
+	private Provider<MenuManager> menuManager;
+
+	@Inject
+	private ClientToolbar clientToolbar;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: This is unused.

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+import net.runelite.client.config.ConfigItem;
+import net.runelite.client.config.Units;
+
+@ConfigGroup(EmojiPlugin.CONFIG_GROUP_KEY)
+public interface EmojiConfig extends Config
+{
+	@ConfigItem(
+		position = 1,
+		keyName = "showEmitEmojiEntry",
+		name = "Add trigger to All chat tab",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java]:
```suggestion
		name = "Add Emoji key menu option to All chat tab",
```

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+		description = "Add trigger to All tab right-click to display possible emojis"
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java]:
```suggestion
		description = "Add menu option to All tab right-click to display a key of available emojis and their trigger text"
```

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+	)
+	default boolean showEmitEmojiEntry()
+	{
+		return false;
+	}
+
+	@ConfigItem(
+		position = 2,
+		keyName = "emojiEmitDuration",
+		name = "Message duration",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java]:
```suggestion
		name = "Emoji key duration",
```

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+		description = "Configure how long the emoji hint message stays in chat"
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiConfig.java]:
```suggestion
		description = "Configure how long the emoji key message is displayed in chat"
```

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 24]
```diff
@@ -222,4 +329,63 @@ private static String removeTags(String str)
...
+		if (removed && !force)
+		{
+			clientThread.invoke(() -> client.runScript(ScriptID.BUILD_CHATBOX));
+		}
+	}
+
+	private String getEmojiMessage()
+	{
+		ChatMessageBuilder builder = new ChatMessageBuilder();
+		ChatMessageBuilder spacer = new ChatMessageBuilder();
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Move this inside the loop and remove the `spacer = new ChatMessageBuilder();` line at the end of the loop

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 25]
```diff
@@ -166,6 +225,54 @@ public void onOverheadTextChanged(final OverheadTextChanged event)
...
 
+
+	@Subscribe
+	public void onGameTick(GameTick gameTick)
+	{
+		timeoutEmojiMessages(false);
+	}
+
+	@Subscribe
+	public void onMenuOptionClicked(MenuOptionClicked event)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: This should subscribe to `WidgetMenuOptionClicked` now that MenuManager is being used to add the menu options. Similarly, this should guard on both the menu option text (like it currently does) and that the widget is one of the all tabs. I would reference other plugins making use of MenuManager such as the inventory tags and instance map plugins to see similar usages there.

Ex:

```java
private static Set<WidgetInfo> ALL_TABS = ImmutableSet.of(
	FIXED_INVENTORY_ALL_TAB,
	RESIZABLE_INVENTORY_ALL_TAB
);

@Subscribe
public void onWidgetMenuOptionClicked(WidgetMenuOptionClicked event)
{
	if (ALL_TABS.contains(event.getWidget())
		&& event.getMenuOption().equals(EMIT)
		&& event.getMenuTarget().equals(MENU_TARGET))
	{
		timeoutEmojiMessages(true);
		addEmojiPaletteMessage();
	}
}
```

Creating an `ALL_TABS` collection would have the benefit of de-duplicating the menu manager calls later on as well.

**Comment by raiyni** [comment]
This needs rebased

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 26]
```diff
@@ -198,28 +330,62 @@ String updateMessage(final String message)
 		return Strings.join(messageWords, " ");
 	}
 
-	/**
-	 * Remove tags, except for &lt;lt&gt; and &lt;gt&gt;
-	 *
-	 * @return
-	 */
-	private static String removeTags(String str)
+	private void timeoutEmojiMessages(boolean force)
```
[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: We still need to remove the current outputted message if the user triggers the menu option again, so I think we still need `currentMessage` field to exist in the class so we can reference it without passing it by params.

If you want I can still use `executorService` to call `timeoutEmojiMessage` after x seconds/ticks but we do need an instance of that class I think.

[the remark this answers is not in the document]
Comment (devLotto) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Yes. The field would still need to exist if you want to do that, but it would just be a MessageNode instead.

**Comment by hjdarnel** [comment]
I rebased and addressed most of the feed. I was having an unescaped while loop in my `getEmojiMessage()` when I used `new StringBuilder` inside my loop, so I need to take another crack at it tomorrow or this weekend. Thanks for your feedback!

File: runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java [hunk 27]
```diff
@@ -198,28 +330,62 @@ String updateMessage(final String message)
...
+			{
+				spacer.append(" ");
+			}
+
+			if (i % EMOJI_PER_LINE == 0)
+			{
+				spacer.append("\n");
+			}
+
+			builder.append(Text.unescapeTags(spacer.build()));
```
[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: I'm not sure what's going on here. You can see I moved the instantiation to inside the loop. When I try to swap to a `StringBuilder` I get a continual loop here.
 ```
while (fontFace.getTextWidth(spacer.build()) < 55)
{
	spacer.append(" ");
}
```

and I can't figure out why. When I compare the value of `fontFace.getTextWidth(spacer.build()` before and after appending some spaces, the width is indeed increasing. Why would this loop forever then?

Otherwise, I've addressed all feedback so far except using `StringBuilder` instead of `ChatMessageBuilder`

[the remark this answers is not in the document]
Comment (hjdarnel) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: Yep, not sure what's going on here. When I use a `StringBuilder#append(String)` to add spaces in this loop, the `FontTypeFace fontFace.getTextWidth(spacer.build())` call isn't increasing in value. Even when testing against a string with length > 100, `getTextWidth()` is returning the same value as though the string is getting trimmed or something. Not sure what to do here.

[the remark this answers is not in the document]
Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/emojis/EmojiPlugin.java]: You should set a breakpoint here and figure out why it only works before making that change.

**Comment by hjdarnel** [comment]
I moved this out into an external plugin to fit what I originally had in mind. https://github.com/runelite/plugin-hub/pull/145

Thank you guys for all your feedback, hate to close this after all that but I like that implementation more.
